install/operator/pkg/syndesis/upgrade/database_upgrade.go

```diff
@@ -0,0 +1,414 @@
+/*
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package upgrade
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"time"
+
+	oappsv1 "github.com/openshift/api/apps/v1"
+	"github.com/spf13/afero"
+	"github.com/syndesisio/syndesis/install/operator/pkg/apis/syndesis/v1beta1"
+	"github.com/syndesisio/syndesis/install/operator/pkg/syndesis/configuration"
+	"github.com/syndesisio/syndesis/install/operator/pkg/util"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const upgradeDeploymentName = "syndesis-db-upgrade"
+
+var upgradeLabels = map[string]string{
+	"syndesis.io/app":       "syndesis",
+	"syndesis.io/component": upgradeDeploymentName,
+}
+
+var upgradeMetadata = metav1.ObjectMeta{
+	Name:   upgradeDeploymentName,
+	Labels: upgradeLabels,
+}
+
+// matches anything followed by space followed by number.number followed (optionally) by another .number and an optional space
+// meant to parse strings like "PostgreSQL 9.5.14" to "9.5" and "postgres (PostgreSQL) 10.6 (Debian 10.6-1.pgdg90+1)" to "10.6"
+var postgresVersionRegex = regexp.MustCompile(`^.* (\d+\.\d+)(?:\.d+)? ?`)
+
+// upgrades the database by leveraging the builtin functionality of the PostgreSQL image from Software Collections
+// key functionality is that the upgrade can be triggered by specifying the `POSTGRESQL_UPGRADE=copy` environment
+// variable
+type databaseUpgrade struct {
+	step
+	syndesis *v1beta1.Syndesis
+	target   func() (float64, error) // target version of PostgreSQL as detected at runtime from the file left by the init container
+	current  func() (float64, error) // current version of PostgreSQL as detected at runtime by querying the running database
+	cleanup  func() error            // how to perform cleanup, that is what to do in case of rollback or when we're done with the upgrade
+}
+
+func newDatabaseUpgrade(base step, s *v1beta1.Syndesis) stepRunner {
```

Comment:
We are implementing pointer semantic for all steps yet this returns a value. Also I think it is clearer if the function return the data type(databaseUpgrade) and not the interface it implements, since it makes it more flexible to use.

Reply:
Actually it's other way around, read about encapsulation and Liskov substitution principle.